pkg/progressbar/progressbar.go

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Ultraviolet
+// SPDX-License-Identifier: Apache-2.0
+package progressbar
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/ultravioletrs/cocos/agent"
+	"golang.org/x/term"
+)
+
+const (
+	progressBarDots = "... "
+	leftBracket     = "["
+	rightBracket    = "]"
+	head            = ">"
+	body            = "="
+	bodyPadding     = "."
+	bufferSize      = 1024 * 1024
+)
+
+type ProgressBar struct {
+	numberOfBytes           int
+	currentUploadedBytes    int
+	currentUploadPercentage int
+	description             string
+	maxWidth                int
+}
+
+func New() *ProgressBar {
+	return &ProgressBar{}
+}
+
+func (p *ProgressBar) SendAlgorithm(description string, buffer *bytes.Buffer, stream *agent.AgentService_AlgoClient) error {
+	p.currentUploadedBytes = 0
+	p.currentUploadPercentage = 0
+	p.numberOfBytes = buffer.Len()
+	p.description = description
+
+	buf := make([]byte, bufferSize)
+
+	for {
+		n, err := buffer.Read(buf)
+		if err == io.EOF {
+			if _, err := io.WriteString(os.Stdout, "\n"); err != nil {
+				return err
+			}
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if p.currentUploadedBytes < p.numberOfBytes {
+			p.currentUploadedBytes = p.currentUploadedBytes + n
+			p.currentUploadPercentage = p.currentUploadedBytes * 100.0 / p.numberOfBytes
+		}
+
+		err = (*stream).Send(&agent.AlgoRequest{Algorithm: buf[:n]})
+		if err != nil {
+			return err
+		}
+
+		if err := p.renderProgressBar(); err != nil {
+			return err
+		}
+	}
+
+	if _, err := (*stream).CloseAndRecv(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *ProgressBar) SendData(description string, buffer *bytes.Buffer, stream *agent.AgentService_DataClient) error {
```

Comment:
There is a lot of repeating code in these two functions. Can you make one common function with parameter algo/data to unify this?

Reply:
@danko-miladinovic consider something like this 
```go
type StreamSender interface {
    Send(interface{}) error
    CloseAndRecv() (interface{}, error)
}

func (p *ProgressBar) sendData(description string, buffer *bytes.Buffer, stream StreamSender, createRequest func([]byte) interface{}) error {
    p.reset(description, buffer.Len())

    buf := make([]byte, bufferSize)

    for {
        n, err := buffer.Read(buf)
        if err == io.EOF {
            if _, err := io.WriteString(os.Stdout, "\n"); err != nil {
                return err
            }
            break
        }
        if err != nil {
            return err
        }

        p.updateProgress(n)

        if err := stream.Send(createRequest(buf[:n])); err != nil {
            return err
        }

        if err := p.renderProgressBar(); err != nil {
            return err
        }
    }

    _, err := stream.CloseAndRecv()
    return err
}

func (p *ProgressBar) reset(description string, totalBytes int) {
    p.currentUploadedBytes = 0
    p.currentUploadPercentage = 0
    p.numberOfBytes = totalBytes
    p.description = description
}

func (p *ProgressBar) updateProgress(bytesRead int) {
    if p.currentUploadedBytes < p.numberOfBytes {
        p.currentUploadedBytes += bytesRead
        p.currentUploadPercentage = float64(p.currentUploadedBytes) * 100.0 / float64(p.numberOfBytes)
    }
}

func (p *ProgressBar) SendAlgorithm(description string, buffer *bytes.Buffer, stream *agent.AgentService_AlgoClient) error {
    return p.sendData(description, buffer, *stream, func(data []byte) interface{} {
        return &agent.AlgoRequest{Algorithm: data}
    })
}

func (p *ProgressBar) SendData(description string, buffer *bytes.Buffer, stream *agent.AgentService_DataClient) error {
    return p.sendData(description, buffer, *stream, func(data []byte) interface{} {
        return &agent.DataRequest{Dataset: data}
    })
}
```